prov/util/src/util_attr.c

```diff
@@ -426,43 +426,76 @@ static int fi_resource_mgmt_level(enum fi_resource_mgmt rm_model)
 /* If a provider supports basic registration mode it should set the FI_MR_BASIC
  * mode bit in prov_mode. Support for FI_MR_SCALABLE is indicated by not setting
  * any of OFI_MR_BASIC_MAP bits. */
-int ofi_check_mr_mode(uint32_t api_version, uint32_t prov_mode,
+int ofi_check_mr_mode(const struct fi_provider *prov, uint32_t api_version,
+		      uint64_t user_info_caps, uint32_t prov_mode,
 		      uint32_t user_mode)
 {
+	uint64_t prov_mode_log;
+	int ret;
+
 	if (FI_VERSION_LT(api_version, FI_VERSION(1, 5))) {
+		if (!(user_info_caps & (FI_RMA | FI_ATOMIC)) &&
```

Comment:
This conditional seems to be making some assumptions and it's non-obvious to me. Is the assumption that if RMA is not used and local registration is not _required_ by the provider, then the `mr_mode` isn't worth checking at all? 

Doesn't the lack of `FI_MR_LOCAL` imply that registration is optional and up to the application to enable? In that case, wouldn't it be wrong to skip checking the `mr_mode` if the application may be using registration?

Reply:
> Is the assumption that if RMA is not used and local registration is not required by the provider, then the mr_mode isn't worth checking at all?

Yes. The MR mode bits FI_MR_ALLOCATED, FI_MR_VIRT_ADDR and FI_MR_PROV_KEY are relevant only for RMA/Atomics. FI_MR_ALLOCATED is relevant only when the provider requires local memory registration. But it is wrapped into FI_MR_BASIC for libfabric version < 1.5. Also, these are the only mode bits that are valid pre-1.5. So it's safe to skip mr_mode checking here.

>Doesn't the lack of FI_MR_LOCAL imply that registration is optional and up to the application to enable? In that case, wouldn't it be wrong to skip checking the mr_mode if the application may be using registration?

If a provider doesn't require FI_MR_LOCAL, the app need not register memory for *local* access. This mode bit is a requirement on the app. If the provider doesn't require it, the app won't have to do any memory registration for *local* access. The app can still advertise *support* for local registration by setting this bit (or FI_MR_BASIC for pre-1.5 apps) in hints. Is there a case where an app *wants* to do memory registration?

Reply:
To be more precise, MR_VIRT_ADDR and MR_PROV_KEY only apply if the endpoint will be the _target_ of an RMA or atomic operation.  We should use ofi_rma_target_allowed(user_caps) as the check.

Reply:
> To be more precise, MR_VIRT_ADDR and MR_PROV_KEY only apply if the endpoint will be the target of an RMA or atomic operation. We should use ofi_rma_target_allowed(user_caps) as the check.

Sounds good. Will update the check.